src/common/lru.c

```diff
@@ -262,38 +272,39 @@ daos_lru_ref_hold(struct daos_lru_cache *lcache, void *key,
 	return rc;
 }
 
-static bool
-lru_flush_cond(struct daos_llink *llink, void *arg)
-{
-	if (llink->ll_ref == 1) /* the last refcount */
-		llink->ll_evicted = 1;
-
-	return llink->ll_evicted;
-}
-
-void
-daos_lru_ref_flush(struct daos_lru_cache *lcache)
-{
-	D_DEBUG(DB_TRACE, "Flush LRU cache: %d > %d\n",
-		lcache->dlc_count, lcache->dlc_csize);
-	daos_lru_cache_evict(lcache, lru_flush_cond, NULL);
-}
+#define LRU_TOO_OLD	15
 
 void
 daos_lru_ref_release(struct daos_lru_cache *lcache, struct daos_llink *llink)
 {
+	struct timespec ts;
+
 	D_ASSERT(lcache != NULL && llink != NULL && llink->ll_ref > 1);
+	D_ASSERT(d_list_empty(&llink->ll_qlink));
 
+	d_gettime(&ts);
 	llink->ll_ref--;
 	if (llink->ll_ref == 1) { /* the last refcount */
-		if (llink->ll_evicted || lcache->dlc_csize == 0) {
-			D_DEBUG(DB_TRACE, "Remove %p from LRU cache\n", llink);
-			/* be freed within hash callback */
-			d_hash_rec_delete_at(&lcache->dlc_htable,
-					     &llink->ll_link);
-			lcache->dlc_count--;
+		if (lcache->dlc_csize == 0)
+			llink->ll_evicted = 1;
+
+		if (llink->ll_evicted) {
+			lru_del_evicted(lcache, llink);
+		} else {
+			llink->ll_accessed = ts.tv_sec;
+			D_ASSERT(d_list_empty(&llink->ll_qlink));
+			d_list_add(&llink->ll_qlink, &lcache->dlc_lru);
 		}
 	}
-	if (lcache->dlc_csize && lcache->dlc_count > lcache->dlc_csize)
-		daos_lru_ref_flush(lcache);
+
+	while (!d_list_empty(&lcache->dlc_lru)) {
+		llink = d_list_entry(lcache->dlc_lru.prev, struct daos_llink,
+				     ll_qlink);
+		if (lcache->dlc_count < lcache->dlc_csize &&
+		    llink->ll_accessed + LRU_TOO_OLD >= ts.tv_sec)
```

Comment:
Are you worrying about overhead of read timestamp? 
The reason we what to evict things is because object cache is per-xstream, so container destroy/close wouldn't evict stale objects from the cache, so these cached objects will consume memory. 
I'm OK either way.

Reply:
Yes, I think this gettime isn't quite necessary. If we want to do time based evict, I think we'd let caller to specify timeout (like the lru size specified by caller), since it's a common lru cache can be used for different things.

It seems to me that container close/destroy will evict all objects belong to the container, see vos_cont_close()->vos_obj_cache_evict().